wp-simple-301-redirects.php

```diff
@@ -420,57 +422,102 @@ function get_redirect_by_index( $index ) {
 			return false;
 		}
 
+		function match($redirect, $url) {
+
+			$do_redirect = false;
+			$destination = $redirect['destination'];
+			$request = $redirect['request'];
+			// check if we should use regex search
+			$wildcard = isset( $redirect['wildcard'] ) ? $redirect['wildcard'] : false;
+
+			if ($wildcard === 'true' && strpos($request,'*') !== false) {
+				// wildcard redirect
+
+				// don't allow people to accidentally lock themselves out of admin
+				if ( strpos($url, '/wp-login') !== 0 && strpos($url, '/wp-admin') !== 0 ) {
+
+					// Turn a glob (well, *'s only), into a valid regular expression
+					$request = preg_quote($request, '/');
+
+					// preg_quote will turn * into \*
+					$request = str_replace('\*','(.*)',$request);
+					$pattern = '/^' . rtrim( $request, '/' ) . '/';
+
+					// destination uses * as a replacement token for the first * match
+					// TODO possily remove this, it's confusing, right?
+					$destination = str_replace('*','$1',$destination);
+
+					// preg_replace will return NULL if the pattern is an error
+					// so best to preg_match first and make sure this is valid
+					// also slightly cheaper on misses to do matches than replacements
+					if( preg_match($pattern, $url) ) {
+						$do_redirect = preg_replace($pattern, $destination, $url);
+					}
+				}
+			}
+			elseif( rtrim( urldecode( $url ), '/' ) == rtrim( $request, '/' ) ) {
+				// simple comparison redirect
+				$do_redirect = $destination;
+			}
+			return $do_redirect;
+		}
+
+		/**
+		 * call recursively to flatten chained 301s
+		 */
+		function resolve($redirects, $request, $safety) {
+			if( $safety > 20 ) {
+				// we're in too deep!
+				return $request;
+			}
+			foreach($redirects as $key => $redirect) {

```

Comment:
So as far as I can tell, the 'glob matching' is making things pretty inefficient.  For any uncached request, we have to do `r` iterations of `match()` in here (where `r` is the total number of redirects we have).

For redirects in a chain, we have to do `r * c` iterations - where `c` is the length of the chain.  I know this all gets varnish cached, but it seems to me like there's a fair bit of needless computation.

Instead, can't we just drop the (dangerous IMO) glob matching feature and just use a hash table lookup?  i.e. O(1).  You could still use the recursion for chains, I guess and you'd even have room to up the `safety` a lot.  Even better would be if the chain got flattened on save rather than on request.

My 2 cents.

Reply:
Wildcards isn't global. It's only set per-URL and won't be used a lot.